// aws-embedded-metrics/lib/Constants.js
"use strict";
/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates.
 * Licensed under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Constants = void 0;
var Constants;
(function (Constants) {
    Constants[Constants["MAX_DIMENSION_SET_SIZE"] = 30] = "MAX_DIMENSION_SET_SIZE";
    Constants[Constants["MAX_DIMENSION_NAME_LENGTH"] = 250] = "MAX_DIMENSION_NAME_LENGTH";
    Constants[Constants["MAX_DIMENSION_VALUE_LENGTH"] = 1024] = "MAX_DIMENSION_VALUE_LENGTH";
    Constants[Constants["MAX_METRIC_NAME_LENGTH"] = 1024] = "MAX_METRIC_NAME_LENGTH";
    Constants[Constants["MAX_NAMESPACE_LENGTH"] = 256] = "MAX_NAMESPACE_LENGTH";
    Constants["VALID_NAMESPACE_REGEX"] = "^(?=.*\\S)[a-zA-Z0-9 ._#:/-]+$";
    Constants["VALID_DIMENSION_REGEX"] = "^[\0-]+$";
    Constants[Constants["MAX_TIMESTAMP_PAST_AGE"] = 1209600000] = "MAX_TIMESTAMP_PAST_AGE";
    Constants[Constants["MAX_TIMESTAMP_FUTURE_AGE"] = 7200000] = "MAX_TIMESTAMP_FUTURE_AGE";
    Constants["DEFAULT_NAMESPACE"] = "aws-embedded-metrics";
    Constants[Constants["MAX_METRICS_PER_EVENT"] = 100] = "MAX_METRICS_PER_EVENT";
    Constants[Constants["MAX_VALUES_PER_METRIC"] = 100] = "MAX_VALUES_PER_METRIC";
    Constants["DEFAULT_AGENT_HOST"] = "0.0.0.0";
    Constants[Constants["DEFAULT_AGENT_PORT"] = 25888] = "DEFAULT_AGENT_PORT";
})(Constants = exports.Constants || (exports.Constants = {}));
