// aws-embedded-metrics/lib/index.js
"use strict";
/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates.
 * Licensed under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Configuration = exports.StorageResolution = exports.Unit = exports.createMetricsLogger = exports.metricScope = exports.AgentSink = exports.LocalSink = exports.MetricsLogger = void 0;
var MetricsLogger_1 = require("./logger/MetricsLogger");
Object.defineProperty(exports, "MetricsLogger", { enumerable: true, get: function () { return MetricsLogger_1.MetricsLogger; } });
var ConsoleSink_1 = require("./sinks/ConsoleSink");
Object.defineProperty(exports, "LocalSink", { enumerable: true, get: function () { return ConsoleSink_1.ConsoleSink; } });
var AgentSink_1 = require("./sinks/AgentSink");
Object.defineProperty(exports, "AgentSink", { enumerable: true, get: function () { return AgentSink_1.AgentSink; } });
var MetricScope_1 = require("./logger/MetricScope");
Object.defineProperty(exports, "metricScope", { enumerable: true, get: function () { return MetricScope_1.metricScope; } });
var MetricsLoggerFactory_1 = require("./logger/MetricsLoggerFactory");
Object.defineProperty(exports, "createMetricsLogger", { enumerable: true, get: function () { return MetricsLoggerFactory_1.createMetricsLogger; } });
var Unit_1 = require("./logger/Unit");
Object.defineProperty(exports, "Unit", { enumerable: true, get: function () { return Unit_1.Unit; } });
var StorageResolution_1 = require("./logger/StorageResolution");
Object.defineProperty(exports, "StorageResolution", { enumerable: true, get: function () { return StorageResolution_1.StorageResolution; } });
const Configuration_1 = __importDefault(require("./config/Configuration"));
exports.Configuration = Configuration_1.default;
